### PR TITLE
Query History: create and map data model to domain model

### PR DESCRIPTION
**Pull request by norascheuch**
To prevent changes in our domain to impact what and how we store the query history directly, we separated both worlds by creating specific data models.

This PR: 
- Creates new interfaces (data models) for local queries and variant analysis query history storage
- Rewrites constructors for the following classes: 
     - LocalQueryInfo
     - CompletedQueryInfo
- Maps new data models to domain model and vice versa

## Checklist

- [ ] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [ ] Issues have been created for any UI or other user-facing changes made by this pull request.
- [ ] _[Maintainers only]_ If this pull request makes user-facing changes that require documentation changes, open a corresponding docs pull request in the [github/codeql](https://github.com/github/codeql/tree/main/docs/codeql/codeql-for-visual-studio-code) repo and add the `ready-for-doc-review` label there.

### CODE REVIEW

File: extensions/ql-vscode/src/query-history/store/query-history-store.ts [hunk 1]
```diff
@@ -8,11 +8,21 @@ import {
...
+import {
+  ALLOWED_QUERY_HISTORY_VERSIONS,
+  QueryHistoryData,
+  QueryHistoryDataItem,
+} from "./query-history-data";
+import { VariantAnalysisHistoryItem } from "../variant-analysis-history-item";
+import { mapLocalQueryDataItemToDomainModel } from "./query-history-data-mapper";
+import { mapLocalQueryInfoToDataModel } from "./query-history-domain-mapper";
+
+// readFromQueryHistoryFile
```
Comment (robertbrignull) [inline extensions/ql-vscode/src/query-history/store/query-history-store.ts:22]: Leftover comments from development that can be removed?

File: extensions/ql-vscode/src/query-history/store/query-history-store.ts [hunk 2]
```diff
@@ -22,71 +32,54 @@ export async function readQueryHistoryFromFile(
...
-            if ("sucessful" in q.completedQuery) {
-              (q.completedQuery as any).successful = (
-                q.completedQuery as any
-              ).sucessful;
-              delete (q.completedQuery as any).sucessful;
-            }
+    // Map the data models to the domain models.
+    const domainModels: QueryHistoryInfo[] = parsedQueries.map((d) => {
+      if (d.t === "variant-analysis") {
+        const query: VariantAnalysisHistoryItem = d;
```
Comment (robertbrignull) [inline extensions/ql-vscode/src/query-history/store/query-history-store.ts:55]: I really like this approach of declaring the type of `query` here, because it means that it enforces that `VariantAnalysisHistoryItem` and `VariantAnalysisDataItem` are consistent. This should stop them diverging before we can implement the data/domain mappers for variant analyses.

We're not breaking the type system here, but we're getting it to work for us to tell us if the types aren't equivalent.

**Review by robertbrignull — COMMENTED** [review]
I have quite a few comments on this PR, but I don't wan't to be discouraging. I think this PR is definitely the right approach and very close. It just needs polish in a few areas. I'm very happy to talk to explain more or help with potential solutions.

**Comment by robertbrignull** [comment]
The CI failures seem to be happening consistently on the `v2.12.5` and nightly CodeQL builds, for both windows and linux. So that makes it look like it's a genuine failure to do with the CodeQL version. I can also reproduce the failure locally. I'd guess my local tests are using the nightly CodeQL build. At least being able to replicate locally will make it easier to track down. I'll take a look and see if I can work it out.

File: extensions/ql-vscode/test/vscode-tests/cli-integration/queries.test.ts [hunk 3]
```diff
@@ -136,16 +136,16 @@ describeWithCodeQL()("Queries", () => {
...
-        ),
+      const queryInfo = await qs.compileAndRunQueryAgainstDatabase(
+        dbItem,
+        await mockInitialQueryInfo(queryUsingExtensionPath),
+        join(tmpDir.name, "mock-storage-path"),
+        progress,
+        token,
       );
 
+      const result = new CompletedQueryInfo(queryInfo.query, queryInfo.result);
```
Comment (robertbrignull) [inline extensions/ql-vscode/test/vscode-tests/cli-integration/queries.test.ts:147]: I think this will fix the tests. We're not passing in all the necessary optional parameters. I'll also comment on the `CompletedQueryInfo` definition because I think we can help make this error less likely for future developers.
```suggestion
      const result = new CompletedQueryInfo(
        queryInfo.query,
        queryInfo.result,
        queryInfo.logFileLocation,
        queryInfo.successful,
        queryInfo.message,
      );
```

File: extensions/ql-vscode/src/query-results.ts [hunk 4]
```diff
@@ -45,45 +45,32 @@ export interface InitialQueryInfo {
...
+  constructor(
+    public query: QueryEvaluationInfo,
+
+    /**
+     * The legacy result. This is only set when loading from the query history.
+     */
+    public result: legacyMessages.EvaluationResult,
+    public logFileLocation?: string,
+    public successful?: boolean,
+    public message?: string,
```
Comment (robertbrignull) [inline extensions/ql-vscode/src/query-results.ts:57]: I think it would be better if we used `T | undefined` for the type instead of using `?` to mark the parameter as optional. At least for all the parameters up to `message` which don't have default values and used to come from the `QueryWithResults` object. This way when you call the constructor it'll force you to provide a value, even if you explicitly write `undefined`.
```suggestion
  constructor(
    public query: QueryEvaluationInfo,

    /**
     * The legacy result. This is only set when loading from the query history.
     */
    public result: legacyMessages.EvaluationResult,
    public logFileLocation: string | undefined,
    public successful: boolean | undefined,
    public message: string | undefined,
```

**Review by robertbrignull — COMMENTED** [review]
I think I've found how to fix the tests. Didn't turn out to be too hard in the end. I'm glad it's not some arcane bug.

It was only failing for some CodeQL CLI versions because those tests only run on versions `>= 2.12.3`. See the `supportsExtensionPacks` method. So it was failing for the latest released version and nightly. The lexicographic sorting in the check runs interface confused me and made me think it was a random old version and nightly 🙈

File: extensions/ql-vscode/src/query-history/store/local-query-history-item.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  isQuickQuery: boolean;
+  isQuickEval: boolean;
+  quickEvalPosition?: PositionData;
+  queryPath: string;
+  databaseInfo: DatabaseInfoData;
+  start: Date;
+  id: string;
+}
+
+export interface DatabaseInfoData {
```
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/src/query-history/store/local-query-history-item.ts:24]: 👍

File: extensions/ql-vscode/src/query-history/store/query-history-data.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+// Contains models and consts for the data we want to store in the query history store.
+// Changes to these models should be done carefully and account for backwards compatibility of data.
+
+import { LocalQueryDataItem } from "./local-query-history-item";
+import { VariantAnalysisDataItem } from "./variant-analysis-data-item";
+
+export const ALLOWED_QUERY_HISTORY_VERSIONS = [1, 2];
```
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/src/query-history/store/query-history-data.ts:7]: Hmmm, I'm a fan of separating types and behaviour, but agree that it's a matter of taste.

File: extensions/ql-vscode/src/query-history/store/query-history-store.ts [hunk 7]
```diff
@@ -22,71 +32,54 @@ export async function readQueryHistoryFromFile(
...
 
-            if (!("successful" in q.completedQuery)) {
-              (q.completedQuery as any).successful =
-                q.completedQuery.result?.resultType === QueryResultType.SUCCESS;
-            }
-          }
-        }
-        return q;
-      });
+      throw Error("Unexpected or corrupted query history file.");
```
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/src/query-history/store/query-history-store.ts:61]: Since q is of type 'any' here we can't access the t-property. However, I can return the whole thing (`d`), that will be more helpful than nothing..

File: extensions/ql-vscode/src/query-history/store/variant-analysis-data-item.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
+import { QueryLanguage } from "../../common/query-language";
+import { QueryStatus } from "../../query-status";
+import {
+  VariantAnalysisFailureReason,
+  VariantAnalysisRepoStatus,
+  VariantAnalysisStatus,
+} from "../../variant-analysis/shared/variant-analysis";
+
+// Data Model for Variant Analysis Query History Items
+// All data points are modelled, except enums.
```
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/src/query-history/store/variant-analysis-data-item.ts:10]: Hm, I couldn't imagine a case where we would need the content of one of the enums be restricted.

File: extensions/ql-vscode/src/query-results.ts [hunk 9]
```diff
@@ -45,45 +45,32 @@ export interface InitialQueryInfo {
...
-    this.result = evaluation.result;
-
-    this.message = evaluation.message;
-    this.successful = evaluation.successful;
-
-    this.sortedResultsInfo = {};
-    this.resultCount = 0;
-  }
+  constructor(
+    public query: QueryEvaluationInfo,
```
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/src/query-results.ts:53]: Oh yes, sounds good!

File: extensions/ql-vscode/src/query-results.ts (hunk 4 above)
Comment (norascheuch) [reply extensions/ql-vscode/src/query-results.ts]: I'm not entirely sure if this is so good.. here is the commit: https://github.com/github/vscode-codeql/pull/2242/commits/d5e3abf10891bda78434d66a264a2f1936e691b4
wdyt?

File: extensions/ql-vscode/src/query-history/store/local-query-history-item.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
+export interface LocalQueryDataItem {
```
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/src/query-history/store/local-query-history-item.ts]: Sounds very good! Did the renames as suggested.

File: extensions/ql-vscode/src/query-results.ts [hunk 11]
```diff
@@ -49,45 +49,32 @@ export interface InitialQueryInfo {
...
-    this.resultCount = 0;
-  }
+  constructor(
+    public readonly query: QueryEvaluationInfo,
+
+    /**
+     * The legacy result. This is only set when loading from the query history.
+     */
+    public readonly result: legacyMessages.EvaluationResult,
+    public readonly logFileLocation?: string,
```
Comment (charisk) [inline extensions/ql-vscode/src/query-results.ts:59]: Should this be 
```ts
public readonly logFileLocation: string | undefined,
```

Like we do in `LocalQueryInfo`?

Same question for other optional members.

**Review by charisk — COMMENTED** [review]
Some initial comments from me. The general direction seems great, and I really like how it's shaping up! We're basically setting patterns we'll be reusing across the codebase so it makes sense to be extra careful/picky. 

In terms of the actual PR, I'm personally finding it hard to review in a way that a) I won't spend too much time reviewing and b) I'll be confident with my review. This is because there are a few different changes in 1 PR. 

One way this could be split is:
- 1 PR that makes the changes to the existing local query history classes/interfaces.
- 1 PR that introduces data models and mappers.
- 1 PR that uses the data models and makes updates to the query history reading/writing.
(Or 1 PR for all the above but with 3 commits)

I know it's a pain, and maybe someone else is better at reviewing bigger PRs than I am, but I personally find it difficult. There's also a mental barrier (think of it as someone giving you a book to read in one sitting instead of a chapter).

**Comment by elenatanasoiu** [comment]
> I know it's a pain, and maybe someone else is better at reviewing bigger PRs than I am, but I personally find it difficult.

I think this is personal preference. People should be allowed to work how they feel comfortable. 

> There's also a mental barrier (think of it as someone giving you a book to read in one sitting instead of a chapter).

On the other hand, it makes it easier to understand the code later since all the context for the initial scaffolding is in the same place so you don't have to go digging for 3-4 PRs with bits of context. 

tl;dr I think both small PRs and large PRs have their pros/cons and it's up to you to choose which ones you're more comfortable with. I'm personally more comfortable with larger PRs and I don't think there's anything wrong with that.

**Comment by charisk** [comment]
> > I know it's a pain, and maybe someone else is better at reviewing bigger PRs than I am, but I personally find it difficult.
> 
> I think this is personal preference. People should be allowed to work how they feel comfortable.
> 
> > There's also a mental barrier (think of it as someone giving you a book to read in one sitting instead of a chapter).
> 
> On the other hand, it makes it easier to understand the code later since all the context for the initial scaffolding is in the same place so you don't have to go digging for 3-4 PRs with bits of context (and sometimes no context). If we're interested in spreading knowledge in the team, having to build a puzzle from multiple pieces of context isn't very inclusive. With a larger PR you make it easier for people not involved in the work to catch up.
> 
> tl;dr I think both small PRs and large PRs have their pros/cons and it's up to you to choose which ones you're more comfortable with. I'm personally more comfortable with larger PRs and I don't think there's anything wrong with that.

I agree there's a personal preference aspect and hopefully I made that clear in my comment. There are at least 2 people involved in a PR, and it would be good to consider both the author of the PR as well as the reviewer(s). In my case, big PRs (that don't have well split commits) are harder to get into and harder to review.

**Comment by elenatanasoiu** [comment]
>  In my case, big PRs (that don't have well split commits) are harder to get into and harder to review.

I agree with having well split commits, but this comes with experience. I would argue with small PRs, it's harder to build this skill.

File: extensions/ql-vscode/src/query-history/store/query-history-store.ts [hunk 12]
```diff
@@ -22,71 +27,55 @@ export async function readQueryHistoryFromFile(
...
-      if (q.t === "variant-analysis") {
-        // the deserializer doesn't know where the remote queries are stored
-        // so we need to assume here that they exist. Later, we check to
-        // see if they exist on disk.
-        return true;
-      }
-      const resultsPath = q.completedQuery?.query.resultsPaths.resultsPath;
-      return !!resultsPath && (await pathExists(resultsPath));
-    });
+    const filteredDomainModels: Promise<QueryHistoryInfo[]> = asyncFilter(
```
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/src/query-history/store/query-history-store.ts:64]: sounds good!

File: extensions/ql-vscode/src/query-history/store/query-history-store.ts [hunk 13]
```diff
@@ -121,13 +110,26 @@ export async function writeQueryHistoryToFile(
     const filteredQueries = queries.filter((q) =>
       q.t === "local" ? q.completedQuery !== undefined : true,
     );
+
+    // map domain model queries to data model
+    const queryHistoryData = filteredQueries.map((q) => {
+      if (q.t === "variant-analysis") {
+        const query: VariantAnalysisHistoryItem = q;
```
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/src/query-history/store/query-history-store.ts:117]: I think it's fine for now (we only have it twice)

[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/src/query-history/store/query-history-store.ts:117]: 👍

File: extensions/ql-vscode/src/run-queries-shared.ts [hunk 14]
```diff
@@ -382,10 +382,10 @@ export class QueryEvaluationInfo extends QueryOutputDir {
 
 export interface QueryWithResults {
   readonly query: QueryEvaluationInfo;
+  readonly result: legacyMessages.EvaluationResult;
```
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/src/run-queries-shared.ts:385]: It's keeping in line with functions where optional parameters should be at the end (I moved this so that I can compare better).
I think it doesn't hurt but might improve readability so I would like to leave it in.

File: extensions/ql-vscode/src/query-history/store/query-history-store.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/src/query-history/store/query-history-store.ts]: Sounds good, extracted it!

File: extensions/ql-vscode/src/query-history/store/query-history-store.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/src/query-history/store/query-history-store.ts:53]: done

File: extensions/ql-vscode/src/query-results.ts (hunk 11 above)
Comment (norascheuch) [reply extensions/ql-vscode/src/query-results.ts:59]: Hm, in this case the original used `?` already, so I would rather keep it as is for now.

File: extensions/ql-vscode/src/query-history/store/variant-analysis-data-item.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/src/query-history/store/variant-analysis-data-item.ts:10]: I just tried to create a new data-enum for a different PR and found out that TS doesn't accept enum types (that are identical but with different names) as assignable to each other.

File: extensions/ql-vscode/src/query-history/store/local-query-history-item.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (robertbrignull) [reply extensions/ql-vscode/src/query-history/store/local-query-history-item.ts]: Nice. I also like the move of putting the `mapQueryHistoryToDomainModels` method into the `data-mapper.ts` file as the only exported function. Keeps things nice and contained.

File: extensions/ql-vscode/src/query-history/store/data-mapper.ts [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+  return queries.map((d) => {
+    if (d.t === "variant-analysis") {
+      const query: VariantAnalysisHistoryItem = d;
+      return query;
+    } else if (d.t === "local") {
+      return mapLocalQueryDataItemToDomainModel(d);
+    }
+
+    throw Error(
+      `Unexpected or corrupted query history file. Unknown query type: ${d}`,
```
Comment (robertbrignull) [inline extensions/ql-vscode/src/query-history/store/data-mapper.ts:31]:
```suggestion
      `Unexpected or corrupted query history file. Unknown query history item: ${JSON.stringify(d)}`,
```
Thanks for updating, though I'm afraid this will print `[Object object]`. We could stringify the item like this. It could be a lot of text, but it would at least include all the info needed to debug it.

File: extensions/ql-vscode/src/query-history/store/variant-analysis-data-item.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (robertbrignull) [reply extensions/ql-vscode/src/query-history/store/variant-analysis-data-item.ts:10]: > I couldn't imagine a case where we would need the content of one of the enums be restricted.

My reasoning here is that if we change the set of values for an enum then we might not be able to deserialize data. As a hypothetical example using real types, imagine if we deleted the `Canceled` value from the `VariantAnalysisStatus` enum. We wouldn't be able to correctly or safely read any serialized data that contained this enum value.

> I just tried to create a new data-enum for a different PR and found out that TS doesn't accept enum types (that are identical but with different names) as assignable to each other.

Hmm, you're right. That's annoying.

However, I think this is fine and I might see what's going on here. This is just a temporary workaround until we write the full data/domain mappers for the variant analysis history item.

We can fix this when we create the proper mappers, because then we can have two identical enum types and then the mappers can convert from one to the other, like we do in [`processApiRepoStatus`](https://github.com/github/vscode-codeql/blob/9015a57732d332e59ccdf961fdaa380c1348b931/extensions/ql-vscode/src/variant-analysis/variant-analysis-processor.ts#L191). Not modelling the enum types is a compromise while we don't have a proper mapper and are just relying on the types being assignable to each other.

Is that understanding correct?

File: extensions/ql-vscode/src/query-history/store/data-mapper.ts (hunk 15 above)
Comment (robertbrignull) [inline extensions/ql-vscode/src/query-history/store/data-mapper.ts:31]:
```suggestion
      `Unexpected or corrupted query history file. Unknown query history item: ${JSON.stringify(d)}`,
```
Thanks for updating, though I'm afraid this will print `[Object object]`. We could stringify the item like this. It could be a lot of text, but it would at least include all the info needed to debug it.

File: extensions/ql-vscode/src/query-history/store/domain-mapper.ts [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+  queries: QueryHistoryInfo[],
+): QueryHistoryDataItem[] {
+  return queries.map((q) => {
+    if (q.t === "variant-analysis") {
+      const query: VariantAnalysisDataItem = q;
+      return query;
+    } else if (q.t === "local") {
+      return mapLocalQueryInfoToDataModel(q);
+    } else {
+      throw new Error("Query type not supported");
```
Comment (robertbrignull) [inline extensions/ql-vscode/src/query-history/store/domain-mapper.ts:24]:
```suggestion
      assertNever(q);
```
I think we would be better off using `assertNever` here because then the type system will enforce that we can't reach this branch. That's a bit safer than having the type system allow it but then failing at runtime. Since for this mapper the values are coming from the application rather than from serialized data, we can safely trust that the type information will be accurate.

File: extensions/ql-vscode/src/query-results.ts (hunk 4 above)
Comment (robertbrignull) [reply extensions/ql-vscode/src/query-results.ts]: I didn't explain too well so there's been a little confusion, but turns out it's all fine in the end.

I initially suggested changing the `CompletedQueryInfo` type rather than the `LocalQueryInfo` type. When constructing a `CompletedQueryInfo` we want to populate all of the fields from the `QueryWithResults`, but then some of the fields may be undefined. For `LocalQueryInfo` most of the arguments of the constructor are not needed at all in normal usage, and we only want them when deserializing query history.

I see that you updated the `CompletedQueryInfo` type anyway in https://github.com/github/vscode-codeql/pull/2242/commits/e3259bd99b4fbb15061548de16f4c78a6cf62230, so that's good. I don't think applying the same change the `LocalQueryInfo` is an improvement, so unless you disagree I'd suggest reverting https://github.com/github/vscode-codeql/commit/d5e3abf10891bda78434d66a264a2f1936e691b4.

Regarding the tests, I think they have been fixed in the background by Dave Bartolomeo's changes. By introducing the `completeThisQuery` function it's removed the call to the `CompletedQueryInfo` constructor from the tests, and it's that constructor call in the tests that wasn't correct and wasn't passing all the necessary arguments.

So I would recommend reverting https://github.com/github/vscode-codeql/commit/d5e3abf10891bda78434d66a264a2f1936e691b4, but then everything looks good 👍 

Does that make more sense?

**Comment by robertbrignull** [comment]
I've taken another look and I've made a couple more minor suggestions but I think this is basically there and very close to merging. @charisk are any of your comments needing changes or responses, or are you also happy?

I agree the PR is quite large and the history is hard to understand. Thankfully I have the context of pairing on it last week. I'm pretty much just reviewing the code as it stands rather than trying to understand every commit in the history.

File: extensions/ql-vscode/src/query-history/store/data-mapper.ts [hunk 17]
```diff
@@ -0,0 +1,81 @@
+import {
```
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/src/query-history/store/data-mapper.ts:1]: I don't mind either way, the current file names are Roberts suggestion. I'm also happy to take a look at it at a later point, for example when repo-states and repo-tasks are also refactored?

File: extensions/ql-vscode/src/query-history/store/data-mapper.ts (hunk 15 above)
Comment (norascheuch) [reply extensions/ql-vscode/src/query-history/store/data-mapper.ts]: 👍 comitted your suggestion

File: extensions/ql-vscode/src/query-results.ts (hunk 4 above)
Comment (norascheuch) [reply extensions/ql-vscode/src/query-results.ts]: Yes, sounds good!

File: extensions/ql-vscode/src/query-history/store/variant-analysis-data-item.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/src/query-history/store/variant-analysis-data-item.ts:10]: Ah, yes that makes sense!

**Comment by charisk** [comment]
> @charisk are any of your comments needing changes or responses, or are you also happy?

I have one unresolved comment around naming of the data models but we can change that in a later PR. 

I've not had a chance to review all the fields etc. so I've not officially approved, but if @robertbrignull has reviewed then please don't wait up!

File: extensions/ql-vscode/src/query-history/store/local-query-data-item.ts [hunk 18]
```diff
@@ -0,0 +1,92 @@
+export interface LocalQueryDataItem {
```
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/src/query-history/store/local-query-data-item.ts:1]: If we want to do this we need to rename `VariantAnalysisDataItem` to `VariantAnalysisHistoryItemDataModel` as well right?

.. 
I have a number of ideas for better naming in my head 🤔 Is it alright if we do this though after repo-states and repo-tasks are done as well? I promise to remember!

File: extensions/ql-vscode/src/query-history/store/domain-mapper.ts (hunk 16 above)
Comment (norascheuch) [reply extensions/ql-vscode/src/query-history/store/domain-mapper.ts]: 👍

**Review by robertbrignull — APPROVED** [review]
